### PR TITLE
Fix dtype issues

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?

Fixes https://github.com/PyTorchLightning/metrics/issues/492
Fixes https://github.com/PyTorchLightning/metrics/issues/484

Reverts some of the changes in https://github.com/PyTorchLightning/metrics/pull/462.
* Removes the `dtype` property. Metrics state can have different dtypes (like any other `nn.Module`) so the property probably does not make sense for all metrics
* Fixes bug with mixed precision training: `half`, `float`, `double` will now have no effect on the dtype of metric states to prevent that metrics gets auto-cast during mix-precision. Instead if the user really wants to change the dtype of all metric states they can now do it using the `set_dtype` method.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@ethanwharris mind have look? ^^

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by ethanwharris — APPROVED** [review]
Awesome, confirmed that this fixes my issue :smiley: thanks @SkafteNicki !

**Review by Borda — APPROVED** [review]

File: torchmetrics/metric.py [hunk 1]
```diff
@@ -413,30 +412,36 @@ def __setstate__(self, state: Dict[str, Any]) -> None:
...
 
-        Works similar to nn.Module.to but also updates the metrics device and dtype properties
+        Normal dtype casting is not supported by this method instead use the `set_dtype` method instead.
         """
-        # there is diff nb vars in PT 1.5
         out = torch._C._nn._parse_to(*args, **kwargs)
-        self._update_properties(device=out[0], dtype=out[1])
-        return super().to(*args, **kwargs)
+        if len(out) == 4:  # pytorch 1.5 and higher
+            device, dtype, non_blocking, convert_to_format = out
```
Comment (ananthsub) [inline torchmetrics/metric.py:427]: what's the minimum version of pytorch torchmetrics supports?

Comment (SkafteNicki) [reply torchmetrics/metric.py:427]: pytorch v1.3

**Comment by Borda** [comment]
@SkafteNicki seems some tests are running for too long with latest versions

File: requirements/test.txt [hunk 2]
```diff
@@ -10,6 +10,7 @@ twine>=3.2
 isort>=5.0
 mypy>=0.790
 phmdoctest>=1.1.1
+PyYAML>=5.1
```
Comment (SkafteNicki) [inline requirements/test.txt:13]:
```suggestion
```

File: tests/helpers/testers.py [hunk 3]
```diff
@@ -154,6 +155,9 @@ def _class_test(
     if check_scriptable:
         torch.jit.script(metric)
 
+    # check that metrics can be parsed with yaml
+    yaml.dump(metric)
+
```
Comment (SkafteNicki) [inline tests/helpers/testers.py:160]:
```suggestion
```

File: tests/helpers/testers.py [hunk 4]
```diff
@@ -20,6 +20,7 @@
 import numpy as np
 import pytest
 import torch
+import yaml
```
Comment (SkafteNicki) [inline tests/helpers/testers.py:23]:
```suggestion
```
